File: sicstus/averbach_1.2.pl
/*

  Seating puzzle in SICStus Prolog.

 
  From Averbach & Chein "Problem Solving Through Recreational Mathematics", 
  page 2, problem 1.2
  
  """
  Ms X, Ms Y, and Ms Z - and American woman, and Englishwoman, and a 
  Frenchwoman, but not neccessarily in that order, were seated around a 
  circular table, playing a game of Hearts. 
  Each passed three cards to the person on her right.
  Ms Y passed three hearts to the American, 
  Ms X passed the queen of spades and two diamonds to the person who
  passed her cards to the Frenchwoman
  
  Who was the American? The Englishwoman? The Frenchwoman?
  """"

  This model gives the following solution
  table: [1, 2, 3]
  [American, English, French]: [1, 3, 2]
 
          1                      American
                       
      3      2               English   French
   



  Model created by Hakan Kjellerstrand, hakank@bonetmail.com
  See also my SICStus Prolog page: http://www.hakank.org/sicstus/

*/

:-use_module(library(clpfd)).
:-use_module(library(lists)).

go :-
        Women = [American,_English,French],
        domain(Women,1,3),

        Table = [X,Y,_Z],
        domain(Table,1,3),

        all_different(Table),
        all_different(Women),

        rightTo(Y, American),
        leftTo(X, French),

        X #= 1, % symmetry breaking

        labeling([],Women),

        Str = [american, english, french],
        ( foreach(Place,Table),
          foreach(P,Placing),
          param(Str,Women) do
              element(Place,Women,WI),
              nth1(WI,Str,P)
        ),
        write(placing:Placing),nl,
        fd_statistics.



% x is right to y
rightTo(X, Y) :-
    X #= Y + 1 ;
    X #= Y - 2. % around the corner


leftTo(X, Y) :- 
    rightTo(Y,X).

File: gecode/averbach_1.2.cpp
/*
  
  Recreational mathematics in Gecode.

  Problem 1.2 from 
  Averbach & Chein "Problem Solving Through Recreational Mathematics",
  page 2:
  """
  Ms X, Ms Y, and Ms Z - and American woman, and Englishwoman, and a 
  Frenchwoman, but not neccessarily in that order, were seated around 
  a circular table, playing a game of Hearts.
  Each passed three cards to the person on her right.
  Ms Y passed three hearts to the American, 
  Ms X passed the queen of spades and two diamonds to the person 
  who passed her cards to the Frenchwoman
 
  Who was the American? The Englishwoman? The Frenchwoman?
  """

  Compare with the following models:
  * MiniZinc: http://www.hakank.org/minizinc/averbach_1.2.mzn
  * SICStus Prolog: http://www.hakank.org/sicstus/averbach_1.2.pl
  * ECLiPSe: http://www.hakank.org/eclipse/averbach_1.2.ecl

  This Gecode model was created by Hakan Kjellerstrand (hakank@gmail.com)
  Also, see my Gecode page: http://www.hakank.org/gecode/ .

*/

#include <gecode/driver.hh>
#include <gecode/int.hh>
#include <gecode/minimodel.hh>

using namespace Gecode;

using std::cout;
using std::endl;

void rightTo(Space& space, IntVar x, IntVar y) {
  rel(space, (x == y + 1) || (x == y -2) );
}

void leftTo(Space& space, IntVar x, IntVar y) {
  rightTo(space, y,x);
}


class Averbach : public Script {
protected:

  const static int n = 5;
  IntVarArray names;
  IntVarArray profession;


  IntVar American;
  IntVar English;
  IntVar French;

  IntVarArray xtable;

public:

  Averbach(const Options& opt) 
    : 
    American(*this, 1, 3),
    English(*this, 1, 3),
    French(*this, 1, 3),
    xtable(*this, 3, 1, 3)
  {

    IntVar 
      X(xtable[0]),
      Y(xtable[1]),
      Z(xtable[2]);

    IntVarArray ytable(*this, 3, 1, 3);
    ytable[0] = American;
    ytable[1] = English;
    ytable[2] = French;
    
    distinct(*this, xtable);
    distinct(*this, ytable);
    
    rightTo(*this, Y, American);
    leftTo(*this, X, French);

    // symmetry breaking
    rel(*this, X == 1);

    // branching
    branch(*this, xtable, INT_VAR_SIZE_MIN(), INT_VAL_MIN());


  }

  // Print solution
  virtual void
  print(std::ostream& os) const {
    os << "xtable: " << xtable << endl;
    os << "American: " << American << endl;
    os << "English: " << English << endl;
    os << "French: " << French << endl;
    os << endl;
  }


  // Constructor for cloning s
  Averbach(bool share, Averbach& s) : Script(share,s) {
    xtable.update(*this, share, s.xtable);
    American.update(*this, share, s.American);
    English.update(*this, share, s.English);
    French.update(*this, share, s.French);

  }

  // Copy during cloning
  virtual Space*
  copy(bool share) {
    return new Averbach(share,*this);
  }
};


int
main(int argc, char* argv[]) {

  Options opt("Averbach");

  opt.solutions(0);
  opt.icl(ICL_BND);

  opt.parse(argc,argv);

  Script::run<Averbach,DFS,Options>(opt);
    
  return 0;
}


